<a href="https://colab.research.google.com/github/J0KERmin/Tutorial/blob/main/ML_1_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

Огляд даних:

In [6]:
import io
import requests

# Посилання на файл
file_url = 'https://drive.google.com/uc?export=download&id=1iJY-MgxD1vl99vrmRcGdhXZFeKcJQNFc'

# Завантаження файлу
response = requests.get(file_url)
file_data = io.BytesIO(response.content)

# Зчитування Excel-файлу
df = pd.read_csv(file_data)

# Перевіримо перші кілька рядків даних для ознайомлення
print(df.head())


          id  vendor_id  pickup_datetime dropoff_datetime  passenger_count  \
0  id2875421          2  3/14/2016 17:24  3/14/2016 17:32                1   
1  id2377394          1   6/12/2016 0:43   6/12/2016 0:54                1   
2  id3858529          2  1/19/2016 11:35  1/19/2016 12:10                1   
3  id3504673          2   4/6/2016 19:32   4/6/2016 19:39                1   
4  id2181028          2  3/26/2016 13:30  3/26/2016 13:38                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  
0                  N  
1                  N  

2. Залежна змінна

In [7]:
# Створимо новий стовпчик trip_duration - тривалість поїздки в секундах
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])
df['dropoff_datetime'] = pd.to_datetime(df['dropoff_datetime'])
df['trip_duration'] = (df['dropoff_datetime'] - df['pickup_datetime']).dt.total_seconds()

In [8]:
# Перевіримо перші кілька рядків даних після додавання нової колонки
print(df.head())

          id  vendor_id     pickup_datetime    dropoff_datetime  \
0  id2875421          2 2016-03-14 17:24:00 2016-03-14 17:32:00   
1  id2377394          1 2016-06-12 00:43:00 2016-06-12 00:54:00   
2  id3858529          2 2016-01-19 11:35:00 2016-01-19 12:10:00   
3  id3504673          2 2016-04-06 19:32:00 2016-04-06 19:39:00   
4  id2181028          2 2016-03-26 13:30:00 2016-03-26 13:38:00   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N          480.0  
1         40.731152         

3. Видалення зайвого

In [9]:
# Видаляємо стовпчик dropoff_datetime з датасету
df.drop(columns=['dropoff_datetime'], inplace=True)

In [10]:
# Перевіримо перші кілька рядків даних після видалення колонки
print(df.head())

          id  vendor_id     pickup_datetime  passenger_count  \
0  id2875421          2 2016-03-14 17:24:00                1   
1  id2377394          1 2016-06-12 00:43:00                1   
2  id3858529          2 2016-01-19 11:35:00                1   
3  id3504673          2 2016-04-06 19:32:00                1   
4  id2181028          2 2016-03-26 13:30:00                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  trip_duration  
0                  N          480.0  
1                  N          660.0  
2                  N         2100.0  


4. Створення нових ознак

In [11]:
# Створимо нові ознаки на основі наявних
# - Час доби, коли почалася поїздка
df['pickup_hour'] = df['pickup_datetime'].dt.hour
# - День тижня, коли почалася поїздка
df['pickup_day_of_week'] = df['pickup_datetime'].dt.dayofweek
# - Чи була поїздка у вихідний день (субота або неділя)
df['is_weekend'] = df['pickup_day_of_week'].isin([5, 6]).astype(int)

In [12]:
# Перевіримо перші кілька рядків даних після додавання нових ознак
print(df.head())

          id  vendor_id     pickup_datetime  passenger_count  \
0  id2875421          2 2016-03-14 17:24:00                1   
1  id2377394          1 2016-06-12 00:43:00                1   
2  id3858529          2 2016-01-19 11:35:00                1   
3  id3504673          2 2016-04-06 19:32:00                1   
4  id2181028          2 2016-03-26 13:30:00                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  trip_duration  pickup_hour  pickup_day_of_week  \
0                  N          480.0           17                   0   
1       

Коли ми готуємо дані для побудови моделі машинного навчання, важливо створити нові ознаки (фічі), які можуть допомогти моделі краще зрозуміти і передбачити залежну змінну. У нашому випадку ми працюємо з даними про поїздки таксі і наша цільова змінна — це тривалість поїздки. Ось як і які нові ознаки ми створили з наявних даних:

1. Час доби, коли почалася поїздка (pickup_hour)

Чому це важливо: Час доби може значно впливати на тривалість поїздки. Наприклад, поїздки в годину пік (ранок або вечір) можуть тривати довше через затори. Виділення години початку поїздки допомагає моделі врахувати ці варіації.

2. День тижня, коли почалася поїздка (pickup_day_of_week)

День тижня також може впливати на тривалість поїздки. Поїздки в будні дні можуть відрізнятися від поїздок у вихідні через різну активність на дорогах. Виділення дня тижня дозволяє моделі врахувати ці відмінності.

3. Чи була поїздка у вихідний день (is_weekend)

Вихідні дні (субота і неділя) можуть мати іншу динаміку руху в порівнянні з буднями. Створення ознаки, яка вказує на те, чи поїздка була у вихідний день, допомагає моделі врахувати цю відмінність. Це бінарна ознака, де 1 означає, що поїздка була у вихідний, і 0 — у будній день.

Збережемо оновлений датасет у новий CSV файл для подальшого використання

In [ ]:
df.to_csv('taxi_dataset_with_features.csv', index=False)

#Лабораторна робота №2: Машинне навчання

In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_squared_log_error
import numpy as np


# Перевірка перших кількох рядків даних
print(df.head())

          id  vendor_id     pickup_datetime  passenger_count  \
0  id2875421          2 2016-03-14 17:24:00                1   
1  id2377394          1 2016-06-12 00:43:00                1   
2  id3858529          2 2016-01-19 11:35:00                1   
3  id3504673          2 2016-04-06 19:32:00                1   
4  id2181028          2 2016-03-26 13:30:00                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  trip_duration  pickup_hour  pickup_day_of_week  \
0                  N          480.0           17                   0   
1       

2. Формування тестової та навчальної вибірок

Розділимо дані на навчальну та тестову вибірки.

In [18]:
# Вибір ознак та цільової змінної
features = ['vendor_id', 'passenger_count', 'pickup_longitude', 'pickup_latitude',
            'dropoff_longitude', 'dropoff_latitude', 'pickup_hour', 'pickup_day_of_week', 'is_weekend']
target = 'trip_duration'

X = df[features]
y = df[target]

# Розділення даних на навчальну та тестову вибірки (80% тренувальні, 20% тестові)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Перевірка розмірів вибірок
print(X_train.shape, X_test.shape)

(838860, 9) (209715, 9)


3. Побудова лінійних моделей

Побудуємо лінійну модель для прогнозування тривалості поїздки.

In [19]:
# Створення та навчання лінійної моделі
model = LinearRegression()
model.fit(X_train, y_train)

# Прогнозування на тестовій вибірці
y_pred = model.predict(X_test)

# Переведення від’ємних прогнозів у нулі
y_pred = np.maximum(y_pred, 0)


4. Оцінка якості моделей за допомогою метрик

Обчислимо метрики MSE, RMSE, MAE та RMSLE.

In [21]:
# Обчислення метрик
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)
rmsle = np.sqrt(mean_squared_log_error(y_test, y_pred))

# Виведення результатів
print(f"MSE: {mse}")
print(f"RMSE: {rmse}")
print(f"MAE: {mae}")
print(f"RMSLE: {rmsle}")


MSE: 9748064.83884706
RMSE: 3122.18911003915
MAE: 581.6020535206754
RMSLE: 0.9070283230932916


*Висновок*

MSE (Mean Squared Error):

MSE є середнім квадратичним відхиленням прогнозів від фактичних значень. Значення 9748064.83884706 вказує на досить високий рівень помилки. Оскільки ця метрика підносить помилки до квадрату, вона дуже чутлива до великих помилок. Це свідчить про те, що деякі прогнози мають значні відхилення від фактичних значень.
RMSE (Root Mean Squared Error):

RMSE є квадратним коренем із MSE і зберігає одиниці виміру цільової змінної. Значення 3122.18911003915 вказує, що середня помилка прогнозу становить близько 3122 секунд (або близько 52 хвилин). Це також вказує на значні відхилення в прогнозах, що може бути неприйнятним у деяких застосуваннях.
MAE (Mean Absolute Error):

MAE є середнім абсолютним відхиленням прогнозів від фактичних значень. Значення 581.6020535206754 означає, що в середньому прогноз моделі відрізняється від фактичного значення на 581 секунду (або приблизно 9.7 хвилин). Це показує середню помилку, яка є меншою, ніж значення RMSE, що свідчить про наявність великих відхилень у деяких прогнозах, які впливають на RMSE.
RMSLE (Root Mean Squared Logarithmic Error):

RMSLE використовується для врахування відносних помилок і менш чутлива до великих відхилень. Значення 0.9070283230932916 вказує на відносно високу логарифмічну помилку. Ця метрика корисна, коли важливо зменшити відносні помилки для невеликих значень, але тут показує, що модель не є оптимальною для таких випадків.


Висновки про якість моделі

Високі значення MSE та RMSE свідчать про наявність великих помилок у прогнозах. Це може бути результатом відсутності важливих ознак або того, що лінійна модель не здатна адекватно захопити складність взаємозв'язків у даних.

MAE показує середню помилку, яка є меншою порівняно з RMSE, що свідчить про наявність кількох великих помилок, які значно впливають на MSE та RMSE.

RMSLE свідчить про те, що модель має проблеми з прогнозуванням як невеликих, так і великих значень тривалості поїздок.